@xen-orchestra/vmware-explorer/esxi.mjs

```diff
@@ -0,0 +1,270 @@
+import { Client } from 'node-vsphere-soap'
+import { dirname } from 'node:path'
+import { EventEmitter } from 'node:events'
+import { strictEqual } from 'node:assert'
+import fetch from 'node-fetch'
+
+import parseVmdk from './parsers/vmdk.mjs'
+import parseVmsd from './parsers/vmsd.mjs'
+import parseVmx from './parsers/vmx.mjs'
+
+const MAX_SCSI = 9
+const MAX_ETHERNET = 9
+
+export default class Esxi extends EventEmitter {
+  #client
+  #cookies
+  #host
+  #user
+  #password
+  #ready = false
+
+  constructor(host, user, password, sslVerify = true) {
+    super()
+    this.#host = host
+    this.#user = user
+    this.#password = password
+    this.#client = new Client(host, user, password, sslVerify)
+    process.on('warning', this.#eatTlsWarning )
+    this.#client.once('ready', () => {
+      process.off('warning', this.#eatTlsWarning )
+      this.#ready = true
+      this.emit('ready')
+    })
+    this.#client.on('error', err => {
+      process.off('warning', this.#eatTlsWarning )
+      console.error({
+        in:'ERROR',
+        code: err.code,
+        message: err.message
+      })
+      this.emit('error', err)
+    })
+  }
+
+  #eatTlsWarning (/* err */){
+     // console.log('yummy', err.code, err.message)
+  }
+  #exec(cmd, args) {
+    strictEqual(this.#ready, true)
+    const client = this.#client
+    return new Promise(function (resolve, reject) {
+      client.once('error', function (error) {
+        client.off('result', resolve)
+        reject(error)
+      })
+      client.runCommand(cmd, args).once('result', function () {
+        client.off('error', reject)
+        resolve(...arguments)
+      })
+    })
+  }
+
+  async download(dataStore, path, range) {
+    strictEqual(this.#ready, true)
+    const url = `https://${this.#host}/folder/${path}?dsName=${dataStore}`
+    const headers = {}
+    if(this.#cookies){
+      headers.cookie= this.#cookies
+    } else {
+      headers.Authorization = 'Basic ' + Buffer.from(this.#user + ':' + this.#password).toString('base64')
+    }
+    if (range) {
+      headers['content-type'] = 'multipart/byteranges'
+      headers.Range = 'bytes=' + range
+    }
+    const res = await fetch(url, {
+      method: 'GET',
+      headers,
+      highWaterMark: 10 * 1024 * 1024,
+    })
+    if (res.status < 200 || res.status >= 300) {
+      const error = new Error(res.status + ' ' + res.statusText + ' ' + url)
+      error.cause = res
+      throw error
+    }
+    if(res.headers.raw()['set-cookie']){
+      this.#cookies = res.headers.raw()['set-cookie']
```

Comment:
Use `set-cookie-parser`, see https://github.com/vatesfr/xen-orchestra/blob/master/packages/xo-server/src/xo-mixins/proxies.mjs#L453

Reply:
I didn't succeed in making it work with cookie parser, I checked the code and it do exactmy 2 authentication : one for the soap query, and one for the download queries

Reply:
Or you can use:

```js
headers.cookie = res.headers.raw()['set-cookie'][0].split(';')[0]
```